packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -0,0 +1,393 @@
+    const { plugins, ...clonedObject } = proposal;
+    return proposal.plugins.safeSnap.safes.map((safe) => ({ ...clonedObject, safe }));
+  });
+
+  // Return only proposals from supported safes.
+  return expandedProposals.filter((proposal) => isSafeSupported(proposal.safe, supportedModules, params.chainId));
+};
+
+// Get all proposals on supported oSnap modules that have not been discarded. Discards are most likely due to disputes,
+// but can also occur on OOv3 upgrades.
+const getUndiscardedProposals = async (
+  supportedModules: SupportedModules,
+  params: MonitoringParams
+): Promise<Array<TransactionsProposedEvent>> => {
+  // Get all proposals for all supported modules.
+  const allProposals = (
+    await Promise.all(
+      Object.keys(supportedModules).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        return runQueryFilter<TransactionsProposedEvent>(og, og.filters.TransactionsProposed(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+      })
+    )
+  ).flat();
+
+  // Get all deleted proposals for all supported modules.
+  const deletedProposals = (
+    await Promise.all(
+      Object.keys(supportedModules).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        return runQueryFilter<ProposalDeletedEvent>(og, og.filters.ProposalDeleted(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+      })
+    )
+  ).flat();
+
+  // Filter out all proposals that have been deleted by matching assertionId. assertionId should be sufficient property
+  // for filtering as it is derived from module address, transaction content and assertion time among other factors.
+  const deletedAssertionIds = deletedProposals.map((deletedProposal) => deletedProposal.args.assertionId);
+  return allProposals.filter((proposal) => !deletedAssertionIds.includes(proposal.args.assertionId));
+};
+
+// Checks if a safeSnap safe from Snapshot proposal is supported by oSnap automation.
+const isSafeSupported = (safe: SafeSnapSafe, supportedModules: SupportedModules, chainId: number): boolean => {
+  for (const ogAddress in supportedModules) {
+    if (isMatchingSafe(safe, chainId, ogAddress)) return true;
+  }
+  return false;
+};
+
+// Filters out all Snapshot proposals that have been proposed on-chain. This is done by matching safe, explanation and
+// proposed transactions.
+const filterPotentialProposals = (
+  supportedProposals: SnapshotProposalExpanded[],
+  onChainProposals: TransactionsProposedEvent[],
+  params: MonitoringParams
+): SnapshotProposalExpanded[] => {
+  return supportedProposals.filter((supportedProposal) => {
+    const matchingOnChainProposals = onChainProposals.filter((onChainProposal) => {
+      // Check if safe and explanation match
+      if (
+        isMatchingSafe(supportedProposal.safe, params.chainId, onChainProposal.address) &&
```

Comment:
Potentially dumb/unrelated question: can two safes "point to" the same OG module?

What happens if I make a fake snapshot space, fake safe, and point it to a real OG module? Then I do a real vote in my fake space and propose it to the OG with a "correct" explanation?